cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i

```diff
@@ -1047,29 +1713,65 @@ services::Status ClassificationTrainBatchKernel<algorithmFPType, method, cpu>::c
-                    par.nClasses, featTypes, indexedFeatures);
-            else if (indexedFeatures.maxNumIndices() <= 65536)
-                s = computeImpl<algorithmFPType, uint16_t, cpu, daal::algorithms::decision_forest::classification::internal::ModelImpl,
-                                TrainBatchTask<algorithmFPType, uint16_t, hist, cpu> >(
-                    pHostApp, x, y, w, *static_cast<daal::algorithms::decision_forest::classification::internal::ModelImpl *>(&m), rd, par,
-                    par.nClasses, featTypes, indexedFeatures);
-            else
-                s = computeImpl<algorithmFPType, dtrees::internal::IndexedFeatures::IndexType, cpu,
-                                daal::algorithms::decision_forest::classification::internal::ModelImpl,
-                                TrainBatchTask<algorithmFPType, dtrees::internal::IndexedFeatures::IndexType, hist, cpu> >(
-                    pHostApp, x, y, w, *static_cast<daal::algorithms::decision_forest::classification::internal::ModelImpl *>(&m), rd, par,
-                    par.nClasses, featTypes, indexedFeatures);
+            if (par.splitter == decision_forest::training::splitter_mode::best)
+            {
+                if (indexedFeatures.maxNumIndices() <= 256)
+                    s = computeImpl<algorithmFPType, uint8_t, cpu, daal::algorithms::decision_forest::classification::internal::ModelImpl,
+                                    TrainBatchTask<algorithmFPType, uint8_t, hist, UnorderedRespHelper<algorithmFPType, cpu>, cpu> >(
+                        pHostApp, x, y, w, *static_cast<daal::algorithms::decision_forest::classification::internal::ModelImpl *>(&m), rd, par,
+                        par.nClasses, featTypes, indexedFeatures);
+                else if (indexedFeatures.maxNumIndices() <= 65536)
+                    s = computeImpl<algorithmFPType, uint16_t, cpu, daal::algorithms::decision_forest::classification::internal::ModelImpl,
+                                    TrainBatchTask<algorithmFPType, uint16_t, hist, UnorderedRespHelper<algorithmFPType, cpu>, cpu> >(
+                        pHostApp, x, y, w, *static_cast<daal::algorithms::decision_forest::classification::internal::ModelImpl *>(&m), rd, par,
+                        par.nClasses, featTypes, indexedFeatures);
+                else
+                    s = computeImpl<algorithmFPType, dtrees::internal::IndexedFeatures::IndexType, cpu,
+                                    daal::algorithms::decision_forest::classification::internal::ModelImpl,
+                                    TrainBatchTask<algorithmFPType, dtrees::internal::IndexedFeatures::IndexType, hist,
+                                                   UnorderedRespHelper<algorithmFPType, cpu>, cpu> >(
+                        pHostApp, x, y, w, *static_cast<daal::algorithms::decision_forest::classification::internal::ModelImpl *>(&m), rd, par,
+                        par.nClasses, featTypes, indexedFeatures);
```

Comment:
Please consider moving this logic to a separate function in order not to duplicate the code.

Reply:
Really good comment, made me realize that regression wasn't properly including a header file too. New function ::_compute templates out the various helpers from ::compute and reduces code size/complexity in recent commits.